xcodeproject/xcodeproj/schemes.go

```diff
@@ -0,0 +1,253 @@
+		isUserSchememanagementFileExist, err := p.isUserSchememanagementFileExist()
+		if err != nil {
+			return nil, err
+		}
+
+		if isUserSchememanagementFileExist {
+			log.TDebugf("Default scheme found")
+			defaultSchemes := p.ReCreateSchemes()
+			return defaultSchemes, nil
+		}
+
+		if isAutocreateSchemesEnabled {
+			log.TDebugf("Autocreating the default scheme")
+			defaultSchemes := p.ReCreateSchemes()
+			return defaultSchemes, nil
+		}
+
+		return nil, fmt.Errorf("no schemes found and the Xcode project's 'Autocreate schemes' option is disabled")
+	}
+
+	log.TDebugf("%d scheme(s) found", len(schemes))
+	return schemes, nil
+}
+
+// Scheme returns the project's scheme by name and the project's absolute path.
+func (p XcodeProj) Scheme(name string) (*xcscheme.Scheme, string, error) {
+	schemes, err := p.Schemes()
+	if err != nil {
+		return nil, "", err
+	}
+
+	normName := norm.NFC.String(name)
+	for _, scheme := range schemes {
+		if norm.NFC.String(scheme.Name) == normName {
+			return &scheme, p.Path, nil
+		}
+	}
+
+	return nil, "", xcscheme.NotFoundError{Scheme: name, Container: p.Name}
+}
+
+func (p XcodeProj) sharedSchemes() ([]xcscheme.Scheme, error) {
+	sharedSchemeFilePaths, err := p.sharedSchemeFilePaths()
+	if err != nil {
+		return nil, err
+	}
+
+	var sharedSchemes []xcscheme.Scheme
+	for _, pth := range sharedSchemeFilePaths {
+		scheme, err := xcscheme.Open(pth)
+		if err != nil {
+			return nil, err
+		}
+
+		sharedSchemes = append(sharedSchemes, scheme)
+	}
+
+	return sharedSchemes, nil
+}
+
+func (p XcodeProj) visibleSchemes() ([]xcscheme.Scheme, error) {
+	sharedSchemes, err := p.sharedSchemes()
+	if err != nil {
+		return nil, err
+	}
+
+	userSchemes, err := p.userSchemes()
+	if err != nil {
+		return nil, err
+	}
+
+	schemes := append(sharedSchemes, userSchemes...)
+	return schemes, nil
+}
+
+func (p XcodeProj) sharedSchemeFilePaths() ([]string, error) {
+	// <project_name>.xcodeproj/xcshareddata/xcschemes/<scheme_name>.xcscheme
+	sharedSchemesDir := filepath.Join(p.Path, "xcshareddata", "xcschemes")
+	return listSchemeFilePaths(sharedSchemesDir)
+}
+
+func (p XcodeProj) userSchemes() ([]xcscheme.Scheme, error) {
+	userSchemeFilePaths, err := p.userSchemeFilePaths()
+	if err != nil {
+		return nil, err
+	}
+
+	var userSchemes []xcscheme.Scheme
+	for _, pth := range userSchemeFilePaths {
+		scheme, err := xcscheme.Open(pth)
+		if err != nil {
+			return nil, err
+		}
+
+		userSchemes = append(userSchemes, scheme)
+	}
+
+	return userSchemes, nil
+}
+
+func (p XcodeProj) userSchemeFilePaths() ([]string, error) {
+	// <project_name>.xcodeproj/xcuserdata/<current_user>.xcuserdatad/xcschemes/<scheme_name>.xcscheme
+	userSchemesDir, err := p.userSchemesDir()
+	if err != nil {
+		return nil, err
+	}
+	return listSchemeFilePaths(userSchemesDir)
+}
+
+func (p XcodeProj) userSchemesDir() (string, error) {
+	// <project_name>.xcodeproj/xcuserdata/<current_user>.xcuserdatad/xcschemes/
+	currentUser, err := user.Current()
```

Comment:
Does it make sense to check for the current user only, if running in a CI env?
I assume in most cases schemes would be recreated if a specific user schema would not be found, so maybe would not break for most people who relied on the previous version that did not consider the current user.

Reply:
I think we should follow how Xcode behaves, to avoid confusion.
I don't think it is too frequent, that you intentionally make a scheme not shared and then customize it.
Also, steps and tools depending on this package might be running locally too (this is probably the minority of uses, but still I think we should consider it), not just in CI.